README.md
https://yihui-he.github.io/blog/bounding-box-regression-with-uncertainty-for-accurate-object-detection

Deprecated. Please see link above


CHANGELOG.md
*In compliance with the [APACHE-2.0](https://opensource.org/licenses/Apache-2.0) license: I declare that this version of the program contains my modifications, which can be seen through the usual "git" mechanism.*  


2022-03  
Contributor(s):  
Yihui He  
>Update README.md  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-11  
Contributor(s):  
Yihui He 何宜晖  
>Update README.md  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2019-03  
Contributor(s):  
Yihui He 何宜晖  
>Update README.md  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2018-10  
Contributor(s):  
Yihui He 何宜晖  
>Update README.md  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2018-09  
Contributor(s):  
yihui-he  
Yihui He 何宜晖  
>arxiv  
>first-commit  
>Update README.md  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 



detectron/utils/boxes.py
# Copyright (c) 2017-present, Facebook, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##############################################################################
#
# Based on:
# --------------------------------------------------------
# Fast/er R-CNN
# Licensed under The MIT License [see LICENSE for details]
# Written by Ross Girshick
# --------------------------------------------------------

"""Box manipulation functions. The internal Detectron box format is
[x1, y1, x2, y2] where (x1, y1) specify the top-left box corner and (x2, y2)
specify the bottom-right box corner. Boxes from external sources, e.g.,
datasets, may be in other formats (such as [x, y, w, h]) and require conversion.

This module uses a convention that may seem strange at first: the width of a box
is computed as x2 - x1 + 1 (likewise for height). The "+ 1" dates back to old
object detection days when the coordinates were integer pixel indices, rather
than floating point coordinates in a subpixel coordinate frame. A box with x2 =
x1 and y2 = y1 was taken to include a single pixel, having a width of 1, and
hence requiring the "+ 1". Now, most datasets will likely provide boxes with
floating point coordinates and the width should be more reasonably computed as
x2 - x1.

In practice, as long as a model is trained and tested with a consistent
convention either decision seems to be ok (at least in our experience on COCO).
Since we have a long history of training models with the "+ 1" convention, we
are reluctant to change it even if our modern tastes prefer not to use it.
"""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import numpy as np

from detectron.core.config import cfg
import detectron.utils.cython_bbox as cython_bbox
import detectron.utils.cython_nms as cython_nms

bbox_overlaps = cython_bbox.bbox_overlaps


def boxes_area(boxes):
    """Compute the area of an array of boxes.
    Args:
	boxes: xyxy
    """
    w = (boxes[:, 2] - boxes[:, 0] + 1)
    h = (boxes[:, 3] - boxes[:, 1] + 1)
    areas = w * h
    assert np.all(areas >= 0), 'Negative areas founds'
    return areas


def unique_boxes(boxes, scale=1.0):
    """Return indices of unique boxes."""
    v = np.array([1, 1e3, 1e6, 1e9])
    hashes = np.round(boxes * scale).dot(v)
    _, index = np.unique(hashes, return_index=True)
    return np.sort(index)


def xywh_to_xyxy(xywh):
    """Convert [x1 y1 w h] box format to [x1 y1 x2 y2] format."""
    if isinstance(xywh, (list, tuple)):
        # Single box given as a list of coordinates
        assert len(xywh) == 4
        x1, y1 = xywh[0], xywh[1]
        x2 = x1 + np.maximum(0., xywh[2] - 1.)
        y2 = y1 + np.maximum(0., xywh[3] - 1.)
        return (x1, y1, x2, y2)
    elif isinstance(xywh, np.ndarray):
        # Multiple boxes given as a 2D ndarray
        return np.hstack(
            (xywh[:, 0:2], xywh[:, 0:2] + np.maximum(0, xywh[:, 2:4] - 1))
        )
    else:
        raise TypeError('Argument xywh must be a list, tuple, or numpy array.')


def xyxy_to_xywh(xyxy):
    """Convert [x1 y1 x2 y2] box format to [x1 y1 w h] format."""
    if isinstance(xyxy, (list, tuple)):
        # Single box given as a list of coordinates
        assert len(xyxy) == 4
        x1, y1 = xyxy[0], xyxy[1]
        w = xyxy[2] - x1 + 1
        h = xyxy[3] - y1 + 1
        return (x1, y1, w, h)
    elif isinstance(xyxy, np.ndarray):
        # Multiple boxes given as a 2D ndarray
        return np.hstack((xyxy[:, 0:2], xyxy[:, 2:4] - xyxy[:, 0:2] + 1))
    else:
        raise TypeError('Argument xyxy must be a list, tuple, or numpy array.')


def filter_small_boxes(boxes, min_size):
    """Keep boxes with width and height both greater than min_size."""
    w = boxes[:, 2] - boxes[:, 0] + 1
    h = boxes[:, 3] - boxes[:, 1] + 1
    keep = np.where((w > min_size) & (h > min_size))[0]
    return keep


def clip_boxes_to_image(boxes, height, width):
    """Clip an array of boxes to an image with the given height and width.
    Args:
	boxes: xyxy
    """
    boxes[:, [0, 2]] = np.minimum(width - 1., np.maximum(0., boxes[:, [0, 2]]))
    boxes[:, [1, 3]] = np.minimum(height - 1., np.maximum(0., boxes[:, [1, 3]]))
    return boxes


def clip_xyxy_to_image(x1, y1, x2, y2, height, width):
    """Clip coordinates to an image with the given height and width."""
    x1 = np.minimum(width - 1., np.maximum(0., x1))
    y1 = np.minimum(height - 1., np.maximum(0., y1))
    x2 = np.minimum(width - 1., np.maximum(0., x2))
    y2 = np.minimum(height - 1., np.maximum(0., y2))
    return x1, y1, x2, y2


def clip_tiled_boxes(boxes, im_shape):
    """Clip boxes to image boundaries. im_shape is [height, width] and boxes
    has shape (N, 4 * num_tiled_boxes).
    Args:
	boxes: xyxy
    """
    assert boxes.shape[1] % 4 == 0, \
        'boxes.shape[1] is {:d}, but must be divisible by 4.'.format(
        boxes.shape[1]
    )
    # x1 >= 0
    boxes[:, 0::4] = np.maximum(np.minimum(boxes[:, 0::4], im_shape[1] - 1), 0)
    # y1 >= 0
    boxes[:, 1::4] = np.maximum(np.minimum(boxes[:, 1::4], im_shape[0] - 1), 0)
    # x2 < im_shape[1]
    boxes[:, 2::4] = np.maximum(np.minimum(boxes[:, 2::4], im_shape[1] - 1), 0)
    # y2 < im_shape[0]
    boxes[:, 3::4] = np.maximum(np.minimum(boxes[:, 3::4], im_shape[0] - 1), 0)
    return boxes


def bbox_transform(boxes, deltas, weights=(1.0, 1.0, 1.0, 1.0)):
    """Forward transform that maps proposal boxes to predicted ground-truth
    boxes using bounding-box regression deltas. See bbox_transform_inv for a
    description of the weights argument.
    Args:
	boxes: xyxy
    """
    if cfg.XYXY:
	return bbox_transform_xyxy(boxes, deltas, weights)
    if boxes.shape[0] == 0:
        return np.zeros((0, deltas.shape[1]), dtype=deltas.dtype)

    boxes = boxes.astype(deltas.dtype, copy=False)

    widths = boxes[:, 2] - boxes[:, 0] + 1.0
    heights = boxes[:, 3] - boxes[:, 1] + 1.0
    ctr_x = boxes[:, 0] + 0.5 * widths
    ctr_y = boxes[:, 1] + 0.5 * heights

    wx, wy, ww, wh = weights
    dx = deltas[:, 0::4] / wx
    dy = deltas[:, 1::4] / wy
    dw = deltas[:, 2::4] / ww
    dh = deltas[:, 3::4] / wh

    # Prevent sending too large values into np.exp()
    dw = np.minimum(dw, cfg.BBOX_XFORM_CLIP)
    dh = np.minimum(dh, cfg.BBOX_XFORM_CLIP)

    pred_ctr_x = dx * widths[:, np.newaxis] + ctr_x[:, np.newaxis]
    pred_ctr_y = dy * heights[:, np.newaxis] + ctr_y[:, np.newaxis]
    pred_w = np.exp(dw) * widths[:, np.newaxis]
    pred_h = np.exp(dh) * heights[:, np.newaxis]

    pred_boxes = np.zeros(deltas.shape, dtype=deltas.dtype)
    # x1
    pred_boxes[:, 0::4] = pred_ctr_x - 0.5 * pred_w
    # y1
    pred_boxes[:, 1::4] = pred_ctr_y - 0.5 * pred_h
    # x2 (note: "- 1" is correct; don't be fooled by the asymmetry)
    pred_boxes[:, 2::4] = pred_ctr_x + 0.5 * pred_w - 1
    # y2 (note: "- 1" is correct; don't be fooled by the asymmetry)
    pred_boxes[:, 3::4] = pred_ctr_y + 0.5 * pred_h - 1

    return pred_boxes

def bbox_transform_xyxy(boxes, deltas, weights=(1.0, 1.0, 1.0, 1.0)):                                          
    if boxes.shape[0] == 0:                                                     
        return np.zeros((0, deltas.shape[1]), dtype=deltas.dtype)               
                                                                                
    boxes = boxes.astype(deltas.dtype, copy=False)                              
                                                                                
    #  0 1 2 3                                                                  
    #  < v > ^                                                                  
                                                                                
    widths = boxes[:, 2] - boxes[:, 0] + 1.0                                    
    heights = boxes[:, 3] - boxes[:, 1] + 1.0                                   
                                                                                
    l = boxes[:, 0]                                                             
    r = boxes[:, 2]                                                             
    d = boxes[:, 1]                                                             
    u = boxes[:, 3]                                                             
                                                                                
    wx, wy, ww, wh = weights
    dl = deltas[:, 0::4] / wx                                                      
    dr = deltas[:, 1::4] / wy                                                      
    dd = deltas[:, 2::4] / wx                                                      
    du = deltas[:, 3::4] / wy                                                      
    # Prevent sending too large values into np.exp()                            
    dl = np.minimum(dl, cfg.BBOX_XFORM_CLIPe)                                    
    dr = np.minimum(dr, cfg.BBOX_XFORM_CLIPe)                                    
    dd = np.minimum(dd, cfg.BBOX_XFORM_CLIPe)                                    
    du = np.minimum(du, cfg.BBOX_XFORM_CLIPe)                                    
                                                                                
    pred_l = dl * widths[:, np.newaxis] + l[:, np.newaxis]                      
    pred_r = dr * widths[:, np.newaxis] + r[:, np.newaxis]                      
    pred_d = dd * heights[:, np.newaxis] + d[:, np.newaxis]                     
    pred_u = du * heights[:, np.newaxis] + u[:, np.newaxis]                     
                                                                                
    pred_boxes = np.zeros(deltas.shape, dtype=deltas.dtype)                     
    # x1                                                                        
    pred_boxes[:, 0::4] = pred_l                                                
    # y1                                                                        
    pred_boxes[:, 1::4] = pred_d                                                
    # x2                                                                        
    pred_boxes[:, 2::4] = pred_r                                                
    # y2                                                                        
    pred_boxes[:, 3::4] = pred_u                                                
                                                                                
    return pred_boxes 

def bbox_std_transform_xyxy(boxes, bbox_epsilon, describ=False):
    #bbox_std = np.exp(bbox_epsilon) 
    bbox_std = 1 / bbox_epsilon**.5
    if describ:
	import pandas as pd
        print((pd.DataFrame(bbox_std.flatten()).describe()))
    boxes = boxes.astype(bbox_std.dtype, copy=False)
    widths = boxes[:, 2::4] - boxes[:, 0::4] + 1.0
    heights = boxes[:, 3::4] - boxes[:, 1::4] + 1.0

    pred = np.zeros(bbox_std.shape, dtype=bbox_std.dtype)
    # x1
    pred[:, 0::4] = bbox_std[:, 0::4] * widths
    # y1
    pred[:, 1::4] = bbox_std[:, 2::4] * heights
    # x2
    pred[:, 2::4] = bbox_std[:, 1::4] * widths 
    # y2
    pred[:, 3::4] = bbox_std[:, 3::4] * heights

    unnorm_pred = np.zeros(bbox_std.shape, dtype=bbox_std.dtype)
    # x1
    unnorm_pred[:, 0::4] = bbox_std[:, 0::4]
    # y1
    unnorm_pred[:, 1::4] = bbox_std[:, 2::4]
    # x2
    unnorm_pred[:, 2::4] = bbox_std[:, 1::4]
    # y2
    unnorm_pred[:, 3::4] = bbox_std[:, 3::4]

    if describ:
        print((pd.DataFrame(pred.flatten()).describe()))
    return pred, unnorm_pred

def bbox_transform_inv(boxes, gt_boxes, weights=(1.0, 1.0, 1.0, 1.0)):
    """Inverse transform that computes target bounding-box regression deltas
    given proposal boxes and ground-truth boxes. The weights argument should be
    a 4-tuple of multiplicative weights that are applied to the regression
    target.

    In older versions of this code (and in py-faster-rcnn), the weights were set
    such that the regression deltas would have unit standard deviation on the
    training dataset. Presently, rather than computing these statistics exactly,
    we use a fixed set of weights (10., 10., 5., 5.) by default. These are
    approximately the weights one would get from COCO using the previous unit
    stdev heuristic.
    Args:
	boxes: [x1 y1 x2 y2]
	gt_boxes: [x1 y1 x2 y2]
    """
    if cfg.XYXY:
	return bbox_transform_inv_xyxy(boxes, gt_boxes, weights)
    ex_widths = boxes[:, 2] - boxes[:, 0] + 1.0
    ex_heights = boxes[:, 3] - boxes[:, 1] + 1.0
    ex_ctr_x = boxes[:, 0] + 0.5 * ex_widths
    ex_ctr_y = boxes[:, 1] + 0.5 * ex_heights

    gt_widths = gt_boxes[:, 2] - gt_boxes[:, 0] + 1.0
    gt_heights = gt_boxes[:, 3] - gt_boxes[:, 1] + 1.0
    gt_ctr_x = gt_boxes[:, 0] + 0.5 * gt_widths
    gt_ctr_y = gt_boxes[:, 1] + 0.5 * gt_heights

    wx, wy, ww, wh = weights
    targets_dx = wx * (gt_ctr_x - ex_ctr_x) / ex_widths
    targets_dy = wy * (gt_ctr_y - ex_ctr_y) / ex_heights
    targets_dw = ww * np.log(gt_widths / ex_widths)
    targets_dh = wh * np.log(gt_heights / ex_heights)

    targets = np.vstack((targets_dx, targets_dy, targets_dw,
                         targets_dh)).transpose()
    return targets

def bbox_transform_inv_xyxy(ex_rois, gt_rois, weights=(1.0, 1.0, 1.0, 1.0)):
    #  0 1 2 3
    #  < v > ^

    gt_l = gt_rois[:, 0]
    gt_r = gt_rois[:, 2]
    gt_d = gt_rois[:, 1]
    gt_u = gt_rois[:, 3]

    ex_l = ex_rois[:, 0]
    ex_r = ex_rois[:, 2]
    ex_d = ex_rois[:, 1]
    ex_u = ex_rois[:, 3]

    ex_widths = ex_r - ex_l + 1.0
    ex_heights = ex_u - ex_d + 1.0

    wx, wy, ww, wh = weights
    targets_dl = wx * (gt_l - ex_l) / ex_widths
    targets_dr = wy * (gt_r - ex_r) / ex_widths
    targets_dd = wx * (gt_d - ex_d) / ex_heights
    targets_du = wy * (gt_u - ex_u) / ex_heights

    targets = np.vstack(
        (targets_dl, targets_dr, targets_dd, targets_du)).transpose()
    return targets

def expand_boxes(boxes, scale):
    """Expand an array of boxes by a given scale.
    Args:
	boxes: xyxy
    """
    w_half = (boxes[:, 2] - boxes[:, 0]) * .5
    h_half = (boxes[:, 3] - boxes[:, 1]) * .5
    x_c = (boxes[:, 2] + boxes[:, 0]) * .5
    y_c = (boxes[:, 3] + boxes[:, 1]) * .5

    w_half *= scale
    h_half *= scale

    boxes_exp = np.zeros(boxes.shape)
    boxes_exp[:, 0] = x_c - w_half
    boxes_exp[:, 2] = x_c + w_half
    boxes_exp[:, 1] = y_c - h_half
    boxes_exp[:, 3] = y_c + h_half

    return boxes_exp


def flip_boxes(boxes, im_width):
    """Flip boxes horizontally.
    Args:
	boxes: xyxy
    """
    boxes_flipped = boxes.copy()
    boxes_flipped[:, 0::4] = im_width - boxes[:, 2::4] - 1
    boxes_flipped[:, 2::4] = im_width - boxes[:, 0::4] - 1
    return boxes_flipped


def aspect_ratio(boxes, aspect_ratio):
    """Perform width-relative aspect ratio transformation.
    Args:
	boxes: xyxy
    """
    boxes_ar = boxes.copy()
    boxes_ar[:, 0::4] = aspect_ratio * boxes[:, 0::4]
    boxes_ar[:, 2::4] = aspect_ratio * boxes[:, 2::4]
    return boxes_ar


def box_voting(top_dets, all_dets, thresh, scoring_method='ID', beta=1.0):
    """Apply bounding-box voting to refine `top_dets` by voting with `all_dets`.
    See: https://arxiv.org/abs/1505.01749. Optional score averaging (not in the
    referenced  paper) can be applied by setting `scoring_method` appropriately.
    """
    # top_dets is [N, 5] each row is [x1 y1 x2 y2, sore]
    # all_dets is [N, 5] each row is [x1 y1 x2 y2, sore]
    top_dets_out = top_dets.copy()
    top_boxes = top_dets[:, :4]
    all_boxes = all_dets[:, :4]
    all_scores = all_dets[:, 4]
    top_to_all_overlaps = bbox_overlaps(top_boxes, all_boxes)
    for k in range(top_dets_out.shape[0]):
        inds_to_vote = np.where(top_to_all_overlaps[k] >= thresh)[0]
        boxes_to_vote = all_boxes[inds_to_vote, :]
        ws = all_scores[inds_to_vote]
	try:
            top_dets_out[k, :4] = np.average(boxes_to_vote, axis=0, weights=ws)
	except:
	    pass
        if scoring_method == 'ID':
            # Identity, nothing to do
            pass
        elif scoring_method == 'TEMP_AVG':
            # Average probabilities (considered as P(detected class) vs.
            # P(not the detected class)) after smoothing with a temperature
            # hyperparameter.
            P = np.vstack((ws, 1.0 - ws))
            P_max = np.max(P, axis=0)
            X = np.log(P / P_max)
            X_exp = np.exp(X / beta)
            P_temp = X_exp / np.sum(X_exp, axis=0)
            P_avg = P_temp[0].mean()
            top_dets_out[k, 4] = P_avg
        elif scoring_method == 'AVG':
            # Combine new probs from overlapping boxes
            top_dets_out[k, 4] = ws.mean()
        elif scoring_method == 'IOU_AVG':
            P = ws
            ws = top_to_all_overlaps[k, inds_to_vote]
            P_avg = np.average(P, weights=ws)
            top_dets_out[k, 4] = P_avg
        elif scoring_method == 'GENERALIZED_AVG':
            P_avg = np.mean(ws**beta)**(1.0 / beta)
            top_dets_out[k, 4] = P_avg
        elif scoring_method == 'QUASI_SUM':
            top_dets_out[k, 4] = ws.sum() / float(len(ws))**beta
        else:
            raise NotImplementedError(
                'Unknown scoring method {}'.format(scoring_method)
            )

    return top_dets_out


def nms(dets, thresh):
    """Apply classic DPM-style greedy NMS."""
    if dets.shape[0] == 0:
        return []
    return cython_nms.nms(dets, thresh)


def soft_nms(
    dets, sigma=0.5, overlap_thresh=0.3, score_thresh=0.001, method='linear'
):
    """Apply the soft NMS algorithm from https://arxiv.org/abs/1704.04503."""
    if dets.shape[0] == 0:
        return dets, []

    methods = {'hard': 0, 'linear': 1, 'gaussian': 2}
    assert method in methods, 'Unknown soft_nms method: {}'.format(method)

    dets, keep = cython_nms.soft_nms(
        np.ascontiguousarray(dets, dtype=np.float32),
        np.float32(sigma),
        np.float32(overlap_thresh),
        np.float32(score_thresh),
        np.uint8(methods[method])
    )
    return dets, keep


detectron/modeling/fast_rcnn_heads.py
# Copyright (c) 2017-present, Facebook, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##############################################################################

"""Various network "heads" for classification and bounding box prediction.

The design is as follows:

... -> RoI ----\                               /-> box cls output -> cls loss
                -> RoIFeatureXform -> box head
... -> Feature /                               \-> box reg output -> reg loss
       Map

The Fast R-CNN head produces a feature representation of the RoI for the purpose
of bounding box classification and regression. The box output module converts
the feature representation into classification and regression predictions.
"""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

from detectron.core.config import cfg
from detectron.utils.c2 import const_fill
from detectron.utils.c2 import gauss_fill
from detectron.utils.net import get_group_gn
import detectron.utils.blob as blob_utils

from caffe2.python import workspace, core


# ---------------------------------------------------------------------------- #
# Fast R-CNN outputs and losses
# ---------------------------------------------------------------------------- #

def add_fast_rcnn_outputs(model, blob_in, dim):
    """Add RoI classification and bounding box regression output ops."""
    # Box classification layer
    model.FC(
        blob_in,
        'cls_score',
        dim,
        model.num_classes,
        weight_init=gauss_fill(0.01),
        bias_init=const_fill(0.0)
    )
    if not model.train:  # == if test
        # Only add softmax when testing; during training the softmax is combined
        # with the label cross entropy loss for numerical stability
        model.Softmax('cls_score', 'cls_prob', engine='CUDNN')
    # Box regression layer
    num_bbox_reg_classes = (
        2 if cfg.MODEL.CLS_AGNOSTIC_BBOX_REG else model.num_classes
    )
    model.FC(
        blob_in,
        'bbox_pred',
        dim,
        num_bbox_reg_classes * 4,
        weight_init=gauss_fill(0.001),
        bias_init=const_fill(0.0)
    )
    if cfg.PRED_STD:
        model.FC(
            blob_in,
            'bbox_pred_std',
            dim,
            num_bbox_reg_classes * 4,
            weight_init=gauss_fill(0.0001),
            bias_init=const_fill(1.0)
        )
        model.net.Abs('bbox_pred_std', 'bbox_pred_std_abs')



def add_fast_rcnn_losses(model):
    """Add losses for RoI classification and bounding box regression."""
    cls_prob, loss_cls = model.net.SoftmaxWithLoss(
        ['cls_score', 'labels_int32'], ['cls_prob', 'loss_cls'],
        scale=model.GetLossScale()
    )
    if cfg.PRED_STD:
        #loss_bbox = model.net.KLLoss(
        #    [
        #        'bbox_pred','bbox_pred_std_abs', 'bbox_targets', 'bbox_inside_weights',
        #        
        #    ],
        #    'loss_bbox',
        #    scale=model.GetLossScale()
        #)
        #stop pred
        model.net.Copy('bbox_pred', 'bbox_pred0')
        #model.net.Copy('bbox_pred_std_abs', 'bbox_pred_std_abs0')
        #model.net.StopGradient('bbox_pred0', 'bbox_pred0')
        #model.net.StopGradient('bbox_pred_std_abs0', 'bbox_pred_std_abs0')
        #model.net.StopGradient('bbox_inside_weights', 'bbox_inside_weights')
        #model.net.StopGradient('bbox_targets', 'bbox_targets')
        ################# bbox_std grad, stop pred
        #log(std)
        model.net.ConstantFill([], 'sigma', value=0.0001, shape=(1,))
        model.net.Add(['bbox_pred_std_abs', 'sigma'], 'bbox_pred_std_abs_')
        model.net.Log('bbox_pred_std_abs_', 'bbox_pred_std_abs_log_')
        model.net.Scale('bbox_pred_std_abs_log_', 'bbox_pred_std_abs_log', scale=-0.5*model.GetLossScale())
        model.net.StopGradient('bbox_outside_weights', 'bbox_outside_weights')
        model.net.Mul(['bbox_pred_std_abs_log', 'bbox_outside_weights'], 'bbox_pred_std_abs_logw')
        model.net.ReduceMean('bbox_pred_std_abs_logw', 'bbox_pred_std_abs_logwr', axes=[0])
        bbox_pred_std_abs_logw_loss = model.net.SumElements(
                'bbox_pred_std_abs_logwr', 'bbox_pred_std_abs_logw_loss')
        #pred0 - y
        model.net.Sub(['bbox_pred0', 'bbox_targets'], 'bbox_in')
        #val = in*(pred0 - u)
        model.net.Mul(['bbox_in', 'bbox_inside_weights'], 'bbox_inw')
        #absval
        model.net.Abs('bbox_inw', 'bbox_l1abs')
        #l12 mask
        model.net.ConstantFill([], 'one', value=1., shape=(1,))
        model.net.GE(['bbox_l1abs', 'one'], 'wl1', broadcast=1)
        model.net.LT(['bbox_l1abs', 'one'], 'wl2', broadcast=1)
        model.net.Cast('wl1', 'wl1f', to=core.DataType.FLOAT) 
        model.net.Cast('wl2', 'wl2f', to=core.DataType.FLOAT)
        #model.net.StopGradient('wl1f', 'wl1f')
        #model.net.StopGradient('wl2f', 'wl2f')
        # val^2
        model.net.Mul(['bbox_inw', 'bbox_inw'], 'bbox_sq')
        # 0.5 val^2
        model.net.Mul(['bbox_sq', 'wl2f'], 'bbox_l2_')
        model.net.Scale('bbox_l2_', 'bbox_l2', scale=0.5)
        # absval - 0.5
        model.net.ConstantFill([], 'half', value=.5, shape=(1,))
        model.net.Sub(['bbox_l1abs', 'half'], 'bbox_l1abs_', broadcast=1)
        model.net.Mul(['bbox_l1abs_', 'wl1f'], 'bbox_l1')
        # sml1 = w * l1 + w*l2
        model.net.Add(['bbox_l1', 'bbox_l2'], 'bbox_inws')
        #alpha * sml1
        model.net.StopGradient('bbox_inws', 'bbox_inws')
        model.net.Mul(['bbox_pred_std_abs', 'bbox_inws'], 'bbox_inws_out')
        model.net.Scale('bbox_inws_out', 'bbox_inws_out', scale=model.GetLossScale())
        model.net.ReduceMean('bbox_inws_out', 'bbox_inws_outr', axes=[0])
        bbox_pred_std_abs_mulw_loss = model.net.SumElements(
                ['bbox_inws_outr'], 'bbox_pred_std_abs_mulw_loss')

        #bbox_pred grad, stop std
        loss_bbox = model.net.SmoothL1Loss(
            [
                'bbox_pred', 'bbox_targets', 'bbox_inside_weights',
                'bbox_pred_std_abs'
            ],
            'loss_bbox',
            scale=model.GetLossScale()
        )
        loss_gradients = blob_utils.get_loss_gradients(model, [loss_cls, loss_bbox, 
            bbox_pred_std_abs_mulw_loss, 
            bbox_pred_std_abs_logw_loss
            ])
        model.AddLosses(['loss_cls', 'loss_bbox', 
            'bbox_pred_std_abs_mulw_loss', 'bbox_pred_std_abs_logw_loss'
            ])
    else:
        loss_bbox = model.net.SmoothL1Loss(
            [
                'bbox_pred', 'bbox_targets', 'bbox_inside_weights',
                'bbox_outside_weights'
            ],
            'loss_bbox',
            scale=model.GetLossScale()
        )

        loss_gradients = blob_utils.get_loss_gradients(model, [loss_cls, loss_bbox])
        model.AddLosses(['loss_cls', 'loss_bbox'])
    model.Accuracy(['cls_prob', 'labels_int32'], 'accuracy_cls')
    model.AddMetrics('accuracy_cls')
    return loss_gradients


# ---------------------------------------------------------------------------- #
# Box heads
# ---------------------------------------------------------------------------- #

def add_roi_2mlp_head(model, blob_in, dim_in, spatial_scale):
    """Add a ReLU MLP with two hidden layers."""
    hidden_dim = cfg.FAST_RCNN.MLP_HEAD_DIM
    roi_size = cfg.FAST_RCNN.ROI_XFORM_RESOLUTION
    roi_feat = model.RoIFeatureTransform(
        blob_in,
        'roi_feat',
        blob_rois='rois',
        method=cfg.FAST_RCNN.ROI_XFORM_METHOD,
        resolution=roi_size,
        sampling_ratio=cfg.FAST_RCNN.ROI_XFORM_SAMPLING_RATIO,
        spatial_scale=spatial_scale
    )
    model.FC(roi_feat, 'fc6', dim_in * roi_size * roi_size, hidden_dim)
    model.Relu('fc6', 'fc6')
    model.FC('fc6', 'fc7', hidden_dim, hidden_dim)
    model.Relu('fc7', 'fc7')
    return 'fc7', hidden_dim


def add_roi_Xconv1fc_head(model, blob_in, dim_in, spatial_scale):
    """Add a X conv + 1fc head, as a reference if not using GroupNorm"""
    hidden_dim = cfg.FAST_RCNN.CONV_HEAD_DIM
    roi_size = cfg.FAST_RCNN.ROI_XFORM_RESOLUTION
    roi_feat = model.RoIFeatureTransform(
        blob_in,
        'roi_feat',
        blob_rois='rois',
        method=cfg.FAST_RCNN.ROI_XFORM_METHOD,
        resolution=roi_size,
        sampling_ratio=cfg.FAST_RCNN.ROI_XFORM_SAMPLING_RATIO,
        spatial_scale=spatial_scale
    )

    current = roi_feat
    for i in range(cfg.FAST_RCNN.NUM_STACKED_CONVS):
        current = model.Conv(
            current, 'head_conv' + str(i + 1), dim_in, hidden_dim, 3,
            stride=1, pad=1,
            weight_init=('MSRAFill', {}),
            bias_init=('ConstantFill', {'value': 0.}),
            no_bias=0)
        current = model.Relu(current, current)
        dim_in = hidden_dim

    fc_dim = cfg.FAST_RCNN.MLP_HEAD_DIM
    model.FC(current, 'fc6', dim_in * roi_size * roi_size, fc_dim)
    model.Relu('fc6', 'fc6')
    return 'fc6', fc_dim


def add_roi_Xconv1fc_gn_head(model, blob_in, dim_in, spatial_scale):
    """Add a X conv + 1fc head, with GroupNorm"""
    hidden_dim = cfg.FAST_RCNN.CONV_HEAD_DIM
    roi_size = cfg.FAST_RCNN.ROI_XFORM_RESOLUTION
    roi_feat = model.RoIFeatureTransform(
        blob_in, 'roi_feat',
        blob_rois='rois',
        method=cfg.FAST_RCNN.ROI_XFORM_METHOD,
        resolution=roi_size,
        sampling_ratio=cfg.FAST_RCNN.ROI_XFORM_SAMPLING_RATIO,
        spatial_scale=spatial_scale
    )

    current = roi_feat
    for i in range(cfg.FAST_RCNN.NUM_STACKED_CONVS):
        current = model.ConvGN(
            current, 'head_conv' + str(i + 1), dim_in, hidden_dim, 3,
            group_gn=get_group_gn(hidden_dim),
            stride=1, pad=1,
            weight_init=('MSRAFill', {}),
            bias_init=('ConstantFill', {'value': 0.}))
        current = model.Relu(current, current)
        dim_in = hidden_dim

    fc_dim = cfg.FAST_RCNN.MLP_HEAD_DIM
    model.FC(current, 'fc6', dim_in * roi_size * roi_size, fc_dim)
    model.Relu('fc6', 'fc6')
    return 'fc6', fc_dim
